### PR TITLE
feat: flightsql datasource

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

File: crates/datasources/src/lance/mod.rs [hunk 1]
```diff
@@ -117,7 +117,7 @@ impl ExecutionPlan for LanceInsertExecPlan {
     }
 
     fn schema(&self) -> SchemaRef {
-        TableProvider::schema(&self.dataset)
+        COUNT_SCHEMA.clone()
```
Comment (universalmind303) [inline crates/datasources/src/lance/mod.rs:120]: Is this unrelated to flightsql?

File: crates/datasources/src/common/query.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
+use std::fmt::Write;
+
+use datafusion::arrow::datatypes::SchemaRef;
+use datafusion::error::DataFusionError;
+use datafusion::logical_expr::Expr;
+
+use super::errors::Result;
+use super::util::{self, Datasource};
+
+pub fn from_exprs(
```
Comment (universalmind303) [inline crates/datasources/src/common/query.rs:10]: we shouldn't use the module to provide context for a function name. 

`from_exprs` has no context without the `query::` part. `query::from_exprs`

```suggestion
pub fn exprs_to_query_string(
```

Comment (universalmind303) [reply crates/datasources/src/common/query.rs:10]: I'd also proceed with caution as not all sql dialects are the same, as we add more expr pushdowns, we'll need to specialize for different dialects.

File: crates/datasources/src/flightsql/insert.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    SendableRecordBatchStream,
+    Statistics,
+};
+use futures::StreamExt;
+use tonic::transport::Channel;
+
+use super::FlightSqlSourceConnectionOptions;
+use crate::common::util::create_count_record_batch;
+
+pub struct ExecPlan {
```
Comment (universalmind303) [inline crates/datasources/src/flightsql/insert.rs:29]:
```suggestion
pub struct FlightSqlExecPlan {
```

File: crates/sqlbuiltins/src/functions/table/flightsql.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+use crate::functions::ConstBuiltinFunction;
+
+#[derive(Debug, Clone, Copy)]
+pub struct ReadFlightSql;
+
+impl ConstBuiltinFunction for ReadFlightSql {
+    const NAME: &'static str = "read_flightsql";
+    const DESCRIPTION: &'static str = "Reads a FlightSQL table";
+    const EXAMPLE: &'static str =
+        "SELECT * FROM read_flightsql('https://localhost:37019', 'database', 'token')";
```
Comment (universalmind303) [inline crates/sqlbuiltins/src/functions/table/flightsql.rs:23]: I don't think `database` and `token` should be specified. We should instead allow users to pass in headers as each flightsql implementation likely has it's own requirements.

```sql
read_flightsql('https://localhost:37019', authorization => '<token>', database => 'database', some_other_header => 'some_value')"
```

File: crates/datasources/src/flightsql/mod.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+#[derive(Debug, Clone)]
+pub struct FlightSqlSourceConnectionOptions {
+    pub uri: String,
+    pub database: String,
+    pub table: String,
+    pub token: String,
+}
+
+pub struct FlightSqlSourceProvider {
```
Comment (universalmind303) [inline crates/datasources/src/flightsql/mod.rs:37]:
```suggestion
pub struct FlightSqlTableProvider {
```

File: crates/datasources/src/lance/mod.rs (hunk 1 above)
Comment (tychoish) [reply crates/datasources/src/lance/mod.rs:120]: yeah, I can pull this into a separate PR, mostly didn't want to lose track of this. 

I think our exec plans in general say that they have the schema of their _inputs_ and not their _outputs_ so the problem is not localized to this case.

File: crates/datasources/src/flightsql/mod.rs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        let catalogs = client.get_catalogs().await?;
+        let schema = Arc::new(
+            client
+                .get_db_schemas(CommandGetDbSchemas {
+                    catalog: Some(catalogs.to_string()),
+                    db_schema_filter_pattern: Some(db),
+                })
+                .await?
+                .try_decode_schema()?,
+        );
```
Comment (universalmind303) [inline crates/datasources/src/flightsql/mod.rs:64]: Does this return the correct schema? This doesn't seem right to me.

File: crates/datasources/src/common/query.rs (hunk 2 above)
Comment (tychoish) [reply crates/datasources/src/common/query.rs:10]: > I'd also proceed with caution as not all sql dialects are the same, as we add more expr pushdowns, we'll need to specialize for different dialects.

the code in question already has specialization for this. 

> we shouldn't use the module to provide context for a function name.

meh, I disagree

- https://github.com/rust-lang/rfcs/blob/master/text/0356-no-module-prefixes.md
- https://github.com/rust-lang/api-guidelines/discussions/66

anyway, this isn't the right place for this conversation.

File: crates/datasources/src/flightsql/mod.rs (hunk 6 above)
Comment (tychoish) [reply crates/datasources/src/flightsql/mod.rs:64]: yeah, probably should be a table thing.

File: crates/sqlbuiltins/src/functions/table/flightsql.rs (hunk 4 above)
Comment (tychoish) [reply crates/sqlbuiltins/src/functions/table/flightsql.rs:23]: to be honest, this was just a prototype; my inclination is that we'll actually make this function be `read_influxdb` and avoid providing generic flightsql cases, for exactly the reasons you have (though I think we can entertain the potential fantasy that flightsql is actually generic.)

File: crates/datasources/src/common/query.rs (hunk 2 above)
Comment (universalmind303) [reply crates/datasources/src/common/query.rs:10]: i think this is slightly different than the examples you provided. 

an error inside `io::Error` is still an error, but more specifically it is an IO error, so any of `IOError`, `io::Error`, `Error` still indicate that it's an error. `io` just provides additional context. same with the `log::LogLevel` example. 

`from_exprs` is completely meaningless without `query::from_exprs`.

**Pull request closed without merging**